File: hello_world.py
"""Hello World module.

A simple introductory Python module that demonstrates best practices
including type hints, Google-style docstrings, error handling, and a
CLI entry point.

Example:
    Run directly::

        $ python hello_world.py
        Hello, World!

    Run with a custom name::

        $ python hello_world.py --name Alice
        Hello, Alice!
"""

import argparse
import sys


def generate_greeting(name: str = "World") -> str:
    """Return a personalised greeting string.

    Args:
        name: The name to greet. Defaults to ``"World"``.

    Returns:
        A greeting string in the form ``"Hello, <name>!"``.

    Raises:
        ValueError: If *name* is an empty string.

    Example:
        >>> generate_greeting()
        'Hello, World!'
        >>> generate_greeting("Alice")
        'Hello, Alice!'
    """
    if not name or not name.strip():
        raise ValueError("name must be a non-empty string")
    return f"Hello, {name.strip()}!"


def print_greeting(name: str = "World") -> None:
    """Print a personalised greeting to stdout.

    This is a thin wrapper around :func:`generate_greeting` that prints
    the result rather than returning it — useful as a top-level helper.

    Args:
        name: The name to greet. Defaults to ``"World"``.

    Example:
        >>> print_greeting()
        Hello, World!
        >>> print_greeting("Bob")
        Hello, Bob!
    """
    print(generate_greeting(name))


def _create_argument_parser() -> argparse.ArgumentParser:
    """Build and return the CLI argument parser.

    Returns:
        A configured :class:`argparse.ArgumentParser` instance.
    """
    parser = argparse.ArgumentParser(
        description="Print a friendly Hello World greeting."
    )
    parser.add_argument(
        "--name",
        default="World",
        help="Name to greet (default: %(default)s)",
    )
    return parser


def run_cli(argv: list[str] | None = None) -> int:
    """Entry point for the hello-world CLI.

    Parses command-line arguments and prints a greeting.

    Args:
        argv: Argument list to parse. Defaults to ``sys.argv[1:]`` when
            ``None``.

    Returns:
        Exit code: ``0`` on success, ``1`` on error.

    Example:
        >>> run_cli(["--name", "World"])
        Hello, World!
        0
    """
    parser = _create_argument_parser()
    args = parser.parse_args(argv)

    try:
        print_greeting(args.name)
    except ValueError as exc:
        print(f"Error: {exc}", file=sys.stderr)
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(run_cli())

File: hello_world.go
// Package main implements a simple Hello World CLI in Go.
//
// This is a direct port of hello_world.py, preserving all behaviour:
//   - Default greeting: "Hello, World!"
//   - Custom name via --name flag
//   - Validation: name must be a non-empty, non-whitespace string
//   - Exit code 0 on success, 1 on error
//
// Usage:
//
//	go run hello_world.go
//	go run hello_world.go --name Alice
//	go run hello_world.go --name ""
package main

import (
	"errors"
	"flag"
	"fmt"
	"os"
	"strings"
)

// generateGreeting returns a personalised greeting string.
//
// It trims surrounding whitespace from name before embedding it.
// An empty or whitespace-only name is rejected with an error.
//
// Parameters:
//
//	name - the name to greet; defaults to "World" when the caller
//	       passes an empty string after trimming is accounted for
//	       by the CLI layer.
//
// Returns:
//
//	A greeting string of the form "Hello, <name>!" and a nil error
//	on success, or an empty string and a non-nil error when name is
//	invalid.
//
// Example:
//
//	greeting, err := generateGreeting("Alice")
//	// greeting == "Hello, Alice!", err == nil
func generateGreeting(name string) (string, error) {
	trimmed := strings.TrimSpace(name)
	if trimmed == "" {
		return "", errors.New("name must be a non-empty string")
	}
	return fmt.Sprintf("Hello, %s!", trimmed), nil
}

// printGreeting writes a personalised greeting to stdout.
//
// It is a thin wrapper around generateGreeting that prints the result
// rather than returning it — useful as a top-level helper.
//
// Parameters:
//
//	name - the name to greet.
//
// Returns:
//
//	An error if name is invalid (empty / whitespace-only), nil otherwise.
func printGreeting(name string) error {
	greeting, err := generateGreeting(name)
	if err != nil {
		return err
	}
	fmt.Println(greeting)
	return nil
}

// createArgumentParser configures and returns a *flag.FlagSet for the CLI.
//
// The returned FlagSet has a single string flag:
//
//	--name   Name to greet (default: "World")
//
// Parameters:
//
//	none
//
// Returns:
//
//	A configured *flag.FlagSet and a pointer to the parsed name value.
func createArgumentParser() (*flag.FlagSet, *string) {
	fs := flag.NewFlagSet("hello-world", flag.ContinueOnError)
	name := fs.String("name", "World", "Name to greet (default: World)")
	return fs, name
}

// runCLI is the entry point for the hello-world CLI.
//
// It parses the supplied argument slice (analogous to sys.argv[1:] in
// Python), prints a greeting, and returns an exit code.
//
// Parameters:
//
//	args - slice of CLI arguments to parse, e.g. os.Args[1:].
//
// Returns:
//
//	0 on success, 1 on any error (invalid flag or empty name).
//
// Example:
//
//	code := runCLI([]string{"--name", "World"})
//	// prints: Hello, World!
//	// code == 0
func runCLI(args []string) int {
	parser, name := createArgumentParser()

	if err := parser.Parse(args); err != nil {
		// flag.ContinueOnError already wrote the error to stderr.
		return 1
	}

	if err := printGreeting(*name); err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		return 1
	}

	return 0
}

func main() {
	os.Exit(runCLI(os.Args[1:]))
}
